pkg/ns/api.go

```diff
@@ -241,87 +241,127 @@ func (p *Provider) CreateFilesystem(params CreateFilesystemParams) error {
-			return err
-		} else if !IsAlreadyExistNefError(err) { // EEXIST code there means the filesystem has dependent clones
-			continue
-		}
-
-		// if here then filesystem has dependent clones (EEXIST error on deletion),
-		// need promote the most recent clone to make the filesystem independent
+		mostRecentError = nil
 
 		snapshots, err := p.GetSnapshots(path, true)
 		if err != nil {
-			l.Warnf("filesystem '%s' has clones, but snapshots list was failed: %s (trying again...)", path, err)
-			continue
+			mostRecentError = fmt.Errorf("failed to get snapshot list: %s", err)
+			break
 		}
 
 		var maxCreationTxg int
 		var mostRecentClone string
 		for _, s := range snapshots {
-			// to get "clones" and "creationTxg" fields that are not presented int the list response
+			// to get "clones" and "creationTxg" fields that are not presented in the list response
 			snapshot, err := p.GetSnapshot(s.Path)
 			if err != nil {
-				l.Warnf(
-					"filesystem '%s' has clones, but getting snapshot '%s' was failed: %s (trying again...)",
-					path,
+				mostRecentError = fmt.Errorf("failed to get '%s' snapshost's info: %s", s.Path, err)
+				break
+			}
+			creationTxg, err := strconv.Atoi(snapshot.CreationTxg)
+			if err != nil {
+				mostRecentError = fmt.Errorf(
+					"snapshot '%s': failed to convert 'creationTxg' value '%s' to integer: %s",
 					s.Path,
+					snapshot.CreationTxg,
 					err,
 				)
-				continue
-			}
-			creationTxg, err := strconv.Atoi(snapshot.CreationTxg)
-			if len(snapshot.Clones) > 0 && creationTxg > maxCreationTxg {
+				break
+			} else if len(snapshot.Clones) > 0 && creationTxg > maxCreationTxg {
 				mostRecentClone = snapshot.Clones[0]
+				maxCreationTxg = creationTxg
 			}
 		}
-
-		if mostRecentClone == "" {
-			l.Warnf("filesystem '%s' is supposted to has clones, but it doesn't, trying again...", path)
+		if mostRecentError != nil {
+			// Failed to determine the most recent clone.
+			// Give another chance (or exit if max attempt count exceeded) if any error happened
+			// while getting each snaphost's information. For example, the snapshot got deleted
+			// right after snapshot list request, but before requesting its information.
 			continue
 		}
 
-		l.Infof("filesystem '%s' has clone '%s', promote it to free up the filesystem...", path, mostRecentClone)
-		err = p.PromoteFilesystem(mostRecentClone)
-		if err != nil {
-			l.Warnf("failed to promote filesystem clone '%s': %v (trying again...)", mostRecentClone, err)
+		if mostRecentClone != "" {
+			err := p.PromoteFilesystem(mostRecentClone)
+			if err != nil {
+				mostRecentError = fmt.Errorf("failed to promote clone '%s': %s", mostRecentClone, err)
+				continue
+			}
+		}
+
+		mostRecentError = p.destroyFilesystem(path, params.DestroySnapshots)
+		if err == nil {
+			return nil
+		} else if !IsAlreadyExistNefError(mostRecentError) { // if EEXIST code - filesystem still has dependent clones
+			return mostRecentError
```

Comment:
Why "return" and not "break" ?

Reply:
The code is supposed to return deletion error from `p.destroyFilesystem()` as it is, to propagate NefError with its code. But, you are right, the current solution does not cover EEXIST on the last step (`i == 2`).  I changed it, also line #341 should be `mostRecentError == nil` - changed this too. Thank you!